[PyTorch] Fix inductor CPU masked() body codegen when result dtype is bool and operator is where

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138275
* #138005
* #137918
* #137917
* #137916
* #137915
* #137914
* #137913
* #137912
* #137911
* #137661
* #137426
* #138744
* #138716
* #138655
* #138542
* __->__ #138486

In this case, it looks like we expect the body to be a VecMask (unify_mask_base_type is called by where()), but we didn't make it a VecMask. Now we do.

Differential Revision: [D64702918](https://our.internmc.facebook.com/intern/diff/D64702918/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

swolchok committed:
[PyTorch] Fix inductor CPU masked() body codegen when result dtype is bool and operator is where

In this case, it looks like we expect the body to be a VecMask (unify_mask_base_type is called by where()), but we don't make it a VecMask.

Differential Revision: [D64702918](https://our.internmc.facebook.com/intern/diff/D64702918/)

[ghstack-poisoned]

swolchok commented:
this should fix one of the inductor failures on #137426 . I need to figure out how to repro this issue without that PR and add a regression test.

swolchok committed:
Update on "[PyTorch] Fix inductor CPU masked() body codegen when result dtype is bool and operator is where"


In this case, it looks like we expect the body to be a VecMask (unify_mask_base_type is called by where()), but we don't make it a VecMask.

Differential Revision: [D64702918](https://our.internmc.facebook.com/intern/diff/D64702918/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

swolchok commented:
> figure out how to repro this issue without that PR and add a regression test.

I could use help adding a test here. The failing input to constant_pad_nd on the diff is tensor size (0, 3, 3) and pad=(1, 1, 1, 1, 1, 1). I brute-forced all sizes in range(3, 10) for dim 1 and 2 on main without finding a repro.

leslie-fang-intel reviewed (commented):
Thanks for the fix. I think the changes in this PR looks good to me. But would like to study more about the failure.
It looks like body code generates a scalar var with bool dtype, and we mistake it as `at::vec::Vectorized` instead of `at::vec::VecMask`. It will be great to have a test case.

swolchok committed:
Update on "[PyTorch] Fix inductor CPU masked() body codegen when result dtype is bool and operator is where"


In this case, it looks like we expect the body to be a VecMask (unify_mask_base_type is called by where()), but we didn't make it a VecMask. Now we do.

Differential Revision: [D64702918](https://our.internmc.facebook.com/intern/diff/D64702918/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

swolchok commented:
>  It will be great to have a test case.

done!

swolchok committed:
Update on "[PyTorch] Fix inductor CPU masked() body codegen when result dtype is bool and operator is where"


In this case, it looks like we expect the body to be a VecMask (unify_mask_base_type is called by where()), but we didn't make it a VecMask. Now we do.

Differential Revision: [D64702918](https://our.internmc.facebook.com/intern/diff/D64702918/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

leslie-fang-intel reviewed (approved):
LGTM. Thanks for the fix.

swolchok committed:
Update on "[PyTorch] Fix inductor CPU masked() body codegen when result dtype is bool and operator is where"


In this case, it looks like we expect the body to be a VecMask (unify_mask_base_type is called by where()), but we didn't make it a VecMask. Now we do.

Differential Revision: [D64702918](https://our.internmc.facebook.com/intern/diff/D64702918/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]
